frontend/src/metabase/data-grid/components/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,200 @@
+import { DndContext, pointerWithin } from "@dnd-kit/core";
+import { restrictToHorizontalAxis } from "@dnd-kit/modifiers";
+import {
+  SortableContext,
+  horizontalListSortingStrategy,
+} from "@dnd-kit/sortable";
+import { flexRender } from "@tanstack/react-table";
+import { type Ref, forwardRef, useCallback, useMemo } from "react";
+
+import { AddColumnButton } from "metabase/data-grid/components/AddColumnButton";
+import { SortableHeader } from "metabase/data-grid/components/SortableHeader";
+import { HEADER_HEIGHT } from "metabase/data-grid/constants";
+import type { DataGridInstance } from "metabase/data-grid/types";
+
+import S from "./DataGrid.module.css";
+
+export type DataGridProps<TData> = DataGridInstance<TData>;
+
+export const DataGrid = forwardRef(function DataGrid<TData>(
+  {
+    table,
+    refs,
+    virtualGrid,
+    measureRoot,
+    columnsReordering,
+    onBodyCellClick,
+    onHeaderCellClick,
+    onAddColumnClick,
+    onScroll,
+  }: DataGridProps<TData>,
+  ref: Ref<HTMLDivElement>,
+) {
+  const {
+    virtualColumns,
+    virtualRows,
+    virtualPaddingLeft,
+    virtualPaddingRight,
+    rowVirtualizer,
+  } = virtualGrid;
+
+  const dndContextProps = useMemo(
+    () => ({
+      collisionDetection: pointerWithin,
+      modifiers: [restrictToHorizontalAxis],
+      ...columnsReordering,
+    }),
+    [columnsReordering],
+  );
+
+  const rowMeasureRef = useCallback(
+    (element: HTMLElement | null) => {
+      rowVirtualizer.measureElement(element);
+    },
+    [rowVirtualizer],
+  );
+
+  const isAddColumnButtonSticky =
```

Comment:
JFYI this variable is recalculated when you scroll the table. Window resize won't update the button. I'm not sure if it's designed in this way or not, so pointing out just in case

https://github.com/user-attachments/assets/a2a45b7b-fe0e-40f7-95c5-bbcaa60ab6c5